### PR TITLE
add read-command subcommand

### PR DESCRIPTION
**Pull request by blu-nitro**
- Adds read-command subcommand that reads ci commands from a json file generated by a GitHub webhook for issue_comment and triggers gitlab ci pipelines for them
- Moves original main command to write subcommand

Todo:
- [x] Add unit tests
- [x] Test with real repositories
- [x] Add possible bot names as argument
- [x] Add bot name as GitLab CI variable sent with pipeline trigger

### CODE REVIEW

**Review by robin-nitrokey — COMMENTED** [review]
Thanks for the PR!  I would prefer to have the webhook parsing in a separate module together with some unit test that parse valid and invalid webhook content.

Generally, I think `@<bot> commands...` is the common format for this kind of commands on GitHub.  AFAIS this would use `!commands...` instead.  Is there a special reason for that?

**Comment by blu-nitro** [comment]
I refactored a few things, but tests are yet to come.

> Generally, I think `@<bot> commands...` is the common format for this kind of commands on GitHub. AFAIS this would use `!commands...` instead. Is there a special reason for that?

Honestly the command trigger was just one i remembered seeing on GitHub lately and it was more of a placeholder for now. I changed it to `@bot commands...` now and you could even have multiple commands in multiple lines like this. But i still wanted to retain the possibility to add a comment with lines that don't start with `@bot`.

File: src/args.rs [hunk 1]
```diff
@@ -1,37 +1,64 @@
...
+pub enum Subcommands {
+    /// Post or update a comment to a GitHub PR.
+    ///
+    /// This command tries to find a pull request that wants to merge the given commit. Then it checks
+    /// if there is a comment that is tagged with the given ID. If yes, the comment is updated with the
+    /// provided text. Otherwise, a new comment is created.
+    ///
+    /// The GITHUB_COMMENT_TOKEN environment variable must be set to a token with the write permission
+    /// for issues or pull requests on the target repository.
+    Write(WriteArgs),
```
Comment (robin-nitrokey) [inline src/args.rs]: Maybe `WriteComment` to be more specific?

File: src/args.rs [hunk 2]
```diff
@@ -1,37 +1,64 @@
...
+    /// Reads commands from github issue_comment webhook and writes them to file
+    ///
+    /// The GITHUB_COMMENT_TOKEN environment variable must be set to a token with the read permission
+    /// for issues or pull requests on the target repository.
+    ReadCommand(ReadCommandArgs),
```
Comment (robin-nitrokey) [inline src/args.rs]: The description is outdated.  As the command not only reads the commands but also submits them to the CI, I think something like `ExecCommands`/`TriggerWebhookCommands`/... would be a better name.

File: src/webhook.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            repo,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const WEBHOOK_DATA: &str = r#"
```
Comment (robin-nitrokey) [inline src/webhook.rs]: Could also be moved to a separate file and [`include_str!`](https://doc.rust-lang.org/std/macro.include_str.html)-ed.

File: src/webhook.rs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+              "events_url": "https://api.github.com/users/namespace/events{/privacy}",
+              "received_events_url": "https://api.github.com/users/namespace/received_events",
+              "type": "User",
+              "site_admin": false
+            }
+          }"#;
+
+    #[test]
+    fn test_parse_webhook() {
+        let webhook = Webhook::parse_from_str(WEBHOOK_DATA).unwrap();
```
Comment (robin-nitrokey) [inline src/webhook.rs:106]: Test for `webhook.comment` is missing.

**Review by robin-nitrokey — COMMENTED** [review]
This feature is very specific to our CI setup so I would rename the tool from `github-comment` to something like `nitrokey-ci`.  What do you think?

File: src/args.rs (hunk 1 above)
Comment (blu-nitro) [reply src/args.rs]: done

File: src/args.rs (hunk 2 above)
Comment (blu-nitro) [reply src/args.rs]: done

File: src/webhook.rs (hunk 3 above)
Comment (blu-nitro) [reply src/webhook.rs]: Like this? I'm not sure that in src is the right place for this test data, or is it?

File: src/webhook.rs (hunk 4 above)
Comment (blu-nitro) [reply src/webhook.rs:106]: done

**Comment by blu-nitro** [comment]
> This feature is very specific to our CI setup so I would rename the tool from `github-comment` to something like `nitrokey-ci`. What do you think?

That is true, it is going beyond just acting on comments now. Although it can be used with any GitLab instance and any repository in just the same way, as long as the repository path is the same on GitHub and GitLab. But a name that represents all of that would get very verbose. So i would stick with `nitrokey-ci` and possibly at a later point explain how to use this tool in other scenarios in a readme.

File: src/webhook.rs (hunk 3 above)
Comment (robin-nitrokey) [reply src/webhook.rs]: Yes, like this.  I would just put it in a `test` or `data` subdirectory.

Comment (blu-nitro) [reply src/webhook.rs]: done

File: src/read_command.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        let atbot = &format!("@{bot} ");
+        text.lines()
+            .filter(|line| line.starts_with(atbot))
+            .flat_map(|s| -> Vec<String> {
+                s.to_string()
+                    .replace(atbot, "")
+                    .split(' ')
+                    .map(|s| s.to_string())
+                    .collect()
+            })
```
Comment (robin-nitrokey) [inline src/read_command.rs]: You can keep this as it is, but I just wanted to mention that there is a cool function [`str::strip_prefix`](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix) that could be useful here.  Also, it is most often not necessary to `collect` inside chained iterators:

```suggestion
        text.lines()
            .flat_map(|line| line.strip_prefix(atbot))
            .flat_map(|s| s.split(' '))
```

The first `flat_map` filters for lines starting with `atbot` and removes it at the same time.  The second `flat_map` replaces the `collect` and produces a single iterator over the commands.

**Review by robin-nitrokey — COMMENTED** [review]
Two final remarks before we can merge:
- With `read_command` being renamed to `trigger_webhook_commands` the `read_command` module and `read` function should be renamed too.
- Wouldn’t it make sense to treat a single line as a single command instead of splitting at spaces?  For example, I could imagine commands with additional arguments like `@bot run-ci slow-job-42`.

File: src/webhook_command_utils.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    parse_commands(&webhook.comment, bots)
+}
+
+fn parse_commands(text: &str, bots: &Vec<String>) -> Vec<BotCommand> {
+    let mut bot_commands: Vec<BotCommand> = Vec::new();
+    for bot in bots {
+        let atbot = &format!("@{bot} ");
+        text.lines()
+            .flat_map(|line| line.strip_prefix(atbot))
+            .flat_map(|line| line.split_once(' '))
```
Comment (robin-nitrokey) [inline src/webhook_command_utils.rs]: I think this misses the case of a command without arguments (should also be added to the tests).

Comment (blu-nitro) [reply src/webhook_command_utils.rs]: Yes i caught that in CI as well

File: src/read_command.rs (hunk 5 above)
Comment (blu-nitro) [reply src/read_command.rs]: This is a lot nicer thank you!
Also yes, as this was supposed to be as general purpose as possible, i think you are right in suggesting to add arguments to commands. This is added now.

**Review by robin-nitrokey — APPROVED** [review]
Thanks for the fixes.  LGTM!

**Pull request merged**